Placeholder canvas does not always propagate HDR content format correctly to the layer.

mattwoodrow opened the pull request:
#### fa73c7ed032b2ea59ee5e02a1b4c7288be9ffd18
<pre>
Placeholder canvas does not always propagate HDR content format correctly to the layer.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295095">https://bugs.webkit.org/show_bug.cgi?id=295095</a>
<a href="https://rdar.apple.com/154478292">rdar://154478292</a>

Reviewed by Mike Wyrzykowski and Kimmo Kinnunen.

OffscreenCanvas can send frames via the AsyncSetLayerContents path instead of a
normal RemoteLayerTreeTransaction, so this path needs to set any properties
related to the buffer (like opaque, and wantsExtendedDynamicRangeContent).

Changes AsyncSetLayerContents to pass a RemoteLayerBackingStoreProperties so
that we&apos;re serializing identical data in both paths. Refactors the UI side a bit
so that this buffer can be applied the same in both cases (which should fix a
potential race where RemoteLayerTreeNode::asyncContentsIdentifier() was only
being check for the transaction path).

Enables generated serializers for RemoteLayerBackingStoreProperties.

CALayer contentsFormat and opaque aren&apos;t used when we set contents explicitly,
so remove these from the LayerProperties serialization. Instead, serialize
&apos;wantsExtendedDynamicRangeContent&apos; as part of the backing store (which could
hopefully be determined from the buffer&apos;s pixel format in the future).

Now that we (correctly) require a value for &apos;opaque&apos; in AsyncSetLayerContents,
add a getter to CanvasRenderingContext for this. Add implementations for
WebGL/WebGPU where this might not be the same as what would be inferred from the
pixel format.

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::isOpaque const):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::isOpaque const):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
(WebCore::PlaceholderRenderingContext::setContentsToLayer):
(WebCore::PlaceholderRenderingContext::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContext::pixelFormat const):
(WebCore::pixelFormatToContentsFormat): Deleted.
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::isOpaque const):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegate::isGraphicsLayerCARemoteAsyncContentsDisplayDelegate const):
(WebCore::GraphicsLayerAsyncContentsDisplayDelegate::setContentsFormat): Deleted.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
(WebCore::convertToContentsFormat):
(WebCore::imageBufferPixelFormatIsOpaque):
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.mm:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegateCocoa::tryCopyToLayer):
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStoreProperties::contentsRenderingResourceIdentifier const):
(WebKit::RemoteLayerBackingStoreProperties::isOpaque const): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStoreProperties::RemoteLayerBackingStoreProperties):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::dumpChangedLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::applyBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setOpaque):
(WebKit::PlatformCALayerRemote::setContentsFormat):

Canonical link: <a href="https://commits.webkit.org/297040@main">https://commits.webkit.org/297040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a58093136d5940ae891837a8009a0b1a246929d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110398 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116423 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112361 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30736 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83946 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113346 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24524 "Found 1 new test failure: fast/filter-image/clipped-filter.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99405 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64388 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109830 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23885 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17545 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60219 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17601 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119214 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37438 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27773 "Found 8 new test failures: accessibility/datetime/input-datetime-local-label-value.html http/tests/media/hls/hls-audio-tracks-language.html http/tests/media/hls/hls-webvtt-default.html http/tests/media/hls/hls-webvtt-style.html http/tests/media/hls/range-request-cross-origin.html http/tests/media/hls/track-in-band-multiple-cues.html http/tests/media/hls/track-webvtt-multitracks.html platform/mac/media/encrypted-media/fps-generateRequest.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92914 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37811 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92737 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23620 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37737 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15471 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33381 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37333 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42804 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36995 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40335 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
```diff
@@ -162,19 +162,30 @@ bool tryCopyToLayer(ImageBuffer& buffer) final
-        m_connection->send(Messages::RemoteLayerTreeDrawingAreaProxy::AsyncSetLayerContents(*m_layerID, WTFMove(*backendHandle), clone->renderingResourceIdentifier()), m_drawingArea.toUInt64());
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
+        bool hasExtendedDynamicRangeContent = m_contentsFormat == ContentsFormat::RGBA16F;
+        RemoteLayerBackingStoreProperties properties(WTFMove(*backendHandle), clone->renderingResourceIdentifier(), m_opaque,  hasExtendedDynamicRangeContent);
```
Two spaces between "m_opaque,  hasExtendedDynamicRangeContent" ?

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
```diff
@@ -185,18 +196,14 @@ void setDestinationLayerID(WebCore::PlatformLayerIdentifier layerID)
     std::optional<ImageBufferBackendHandle> m_surfaceBackendHandle WTF_GUARDED_BY_LOCK(m_surfaceLock);
     Markable<WebCore::RenderingResourceIdentifier> m_surfaceIdentifier WTF_GUARDED_BY_LOCK(m_surfaceLock);
     ContentsFormat m_contentsFormat { ContentsFormat::RGBA8 };
+    bool m_opaque;
```
could initialise this just for style and add the guarded_by for this and m_contentsFormat

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
```diff
@@ -803,7 +802,6 @@ static inline bool isEquivalentLayer(const PlatformCALayer* layer, const std::op
         return;
 
     m_properties.contentsFormat = contentsFormat;
```
a bit risky, but I think this function lacks the updateBackingStore(). Not necessarily part of this change, but otoh could be.

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -227,22 +227,26 @@ class RemoteLayerBackingStoreProperties {
 public:
     RemoteLayerBackingStoreProperties() = default;
     RemoteLayerBackingStoreProperties(RemoteLayerBackingStoreProperties&&) = default;
+#if HAVE(SUPPORT_HDR_DISPLAY)
```
I'd be fine without the ifdef, too. I think the use-site uses different ifdef between the constructors, so it's a bit error-prone.

Consider this more of a reminder to check if the headroom is somehow forgotten, not a requirement:

One way to think of it is that :
-  have one constructor for SDR content 
- have one constructor for EDR content, where you also force the client to provide the max requested headroom?

I think currently at least for 2d contexts we could provide the painted headroom, but in this patch you don't propagate it the same way as with layers? I'm not sure it makes sense in 2d canvas incremental rendering style.

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
```diff
@@ -162,19 +162,30 @@ bool tryCopyToLayer(ImageBuffer& buffer) final
         }
 
-        m_connection->send(Messages::RemoteLayerTreeDrawingAreaProxy::AsyncSetLayerContents(*m_layerID, WTFMove(*backendHandle), clone->renderingResourceIdentifier()), m_drawingArea.toUInt64());
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
```
Here's the use-site that is incorrect.
Would fail compile if ENABLE_PIXEL_FORMAT_RGBA16F = false but SUPPORT_HDR... = trye

kkinnunen-apple commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -130,6 +130,24 @@
+        setAsyncContentsIdentifier(*identifier);
+}
+
+
```
one newline too many!!

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -264,6 +268,7 @@ class RemoteLayerBackingStoreProperties {
     RemoteLayerBackingStore::Type m_type;
 #if HAVE(SUPPORT_HDR_DISPLAY)
+    bool m_hasExtendedDynamicRange;
     float m_maxRequestedEDRHeadroom;
```
Might want to initialise the m_maxRequestedEDRHeadRoom to 1.
Also: I cannot find the place that writes RemoteLayeTBackingStore::m_maxRequestedEDRHeadRoom in the WCP side??

mattwoodrow committed:
Placeholder canvas does not always propagate HDR content format correctly to the layer.
https://bugs.webkit.org/show_bug.cgi?id=295095
rdar://154478292

Reviewed by Mike Wyrzykowski and Kimmo Kinnunen.

OffscreenCanvas can send frames via the AsyncSetLayerContents path instead of a
normal RemoteLayerTreeTransaction, so this path needs to set any properties
related to the buffer (like opaque, and wantsExtendedDynamicRangeContent).

Changes AsyncSetLayerContents to pass a RemoteLayerBackingStoreProperties so
that we're serializing identical data in both paths. Refactors the UI side a bit
so that this buffer can be applied the same in both cases (which should fix a
potential race where RemoteLayerTreeNode::asyncContentsIdentifier() was only
being check for the transaction path).

Enables generated serializers for RemoteLayerBackingStoreProperties.

CALayer contentsFormat and opaque aren't used when we set contents explicitly,
so remove these from the LayerProperties serialization. Instead, serialize
'wantsExtendedDynamicRangeContent' as part of the backing store (which could
hopefully be determined from the buffer's pixel format in the future).

Now that we (correctly) require a value for 'opaque' in AsyncSetLayerContents,
add a getter to CanvasRenderingContext for this. Add implementations for
WebGL/WebGPU where this might not be the same as what would be inferred from the
pixel format.

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::isOpaque const):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::isOpaque const):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
(WebCore::PlaceholderRenderingContext::setContentsToLayer):
(WebCore::PlaceholderRenderingContext::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContext::pixelFormat const):
(WebCore::pixelFormatToContentsFormat): Deleted.
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::isOpaque const):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegate::isGraphicsLayerCARemoteAsyncContentsDisplayDelegate const):
(WebCore::GraphicsLayerAsyncContentsDisplayDelegate::setContentsFormat): Deleted.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
(WebCore::convertToContentsFormat):
(WebCore::imageBufferPixelFormatIsOpaque):
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.mm:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegateCocoa::tryCopyToLayer):
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStoreProperties::contentsRenderingResourceIdentifier const):
(WebKit::RemoteLayerBackingStoreProperties::isOpaque const): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStoreProperties::RemoteLayerBackingStoreProperties):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::dumpChangedLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::applyBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setOpaque):
(WebKit::PlatformCALayerRemote::setContentsFormat):

Canonical link: https://commits.webkit.org/297040@main
